cabbage/task_worker.py

```diff
@@ -11,72 +10,83 @@
+            try:
+                logger.debug(f"""About to run task from row {task_row})""")
+                self.run_task(task_row=task_row)
+                status = "done"
+            except exceptions.TaskError:
+                pass
+            finally:
+                logger.debug(f"Calling finish_task({task_id}, {status})")
+                self._task_manager.finish_task(task_row, status=status)
+
+            if self._stop_requested:
+                break
+
+    def run_task(self, task_row: postgres.TaskRow) -> None:
+        task_name = task_row.task_type
+        try:
+            task = self._task_manager.tasks[task_name]
+        except KeyError:
+            raise exceptions.TaskNotFound(task_row)
 
-    for task_row in postgres.get_tasks(cursor=curs, queue=queue):  # pragma: no branch
+        pk = task_row.id
 
-        assert isinstance(task_row.id, int)
-        task_id = task_row.id
+        task_run = tasks.TaskRun(task=task, id=pk, lock=task_row.targeted_object)
+        kwargs = task_row.args
 
-        status = "error"
+        description = f"{task.queue}.{task.name}.{pk}({kwargs})"
+        logger.info(f"Start - {description}")
         try:
-            logger.debug(f"""About to run task from row {task_row})""")
-            call_task(task_manager=task_manager, task_row=task_row)
-            status = "done"
-        except exceptions.TaskError:
-            pass
-        finally:
-            logger.debug(f"Calling finish_task({task_id}, {status})")
-            postgres.finish_task(cursor=curs, task_id=task_id, status=status)
-
-
-def call_task(task_manager: tasks.TaskManager, task_row: postgres.TaskRow) -> None:
-    task_name = task_row.task_type
-    try:
-        task = task_manager.tasks[task_name]
-    except KeyError:
-        raise exceptions.TaskNotFound(task_row)
-
-    pk = task_row.id
-
-    task_run = tasks.TaskRun(task=task, id=pk, lock=task_row.targeted_object)
-    kwargs = task_row.args
-
-    description = f"{task.queue}.{task.name}.{pk}({kwargs})"
-    logger.info(f"Start - {description}")
-    try:
-        task_run.run(**kwargs)
-    except Exception as e:
-        logger.exception(f"Error - {description}")
-        raise exceptions.TaskError() from e
-    else:
-        logger.info(f"Success - {description}")
+            task_run.run(**kwargs)
+        except Exception as e:
+            logger.exception(f"Error - {description}")
+            raise exceptions.TaskError() from e
+        else:
+            logger.info(f"Success - {description}")
+
+    def _install_signal_handlers(self) -> None:
+        def stop(_signum, _frame):
+            self._stop_requested = True
+            logger.info("Stop requested, waiting for task to finish")
+
+        signal.signal(signal.SIGINT, stop)
+        signal.signal(signal.SIGTERM, stop)
+
+    def _uninstall_signal_handlers(self) -> None:
+        signal.signal(signal.SIGINT, signal.SIG_DFL)
+        signal.signal(signal.SIGTERM, signal.SIG_DFL)
```

Comment:
I'd rather signal specific logic be implemented in a context manager, and we make sure we reinstall previously installed handlers when leaving.

For example, I bet we've destroyed pytests signal handling.

Reply:
`signal.getsignal` can give you the info on what's currently there.